### PR TITLE
Upgrade to Python 3

### LINKED ISSUE #18: Is it possible to update mumo to python 3?
I know mumo just works fine, but Python 2 is going to die very soon.

And another thing, it's really annoying to write non-ascii strings in Python 2 (either `u'string'` or import future)

I have tried to update it myself but turns out that is too complicated for me.

### PR DESCRIPTION
**Pull request by an0nfunc**
This is based on work done 5 years back, so testing and reviews are most welcome. Fixes #18.

Tested to be working:

- [x] Basic functionality
- [x] Module 'seen'
- [x] Module 'source'
- [x] Module 'bf2'
- [x] Module 'idlemove'
- [x] Module 'onjoin'
- [x] Module 'samplecontext'
- [x] testsuite.py
- [x] config_test.py
- [x] users_test.py
- [x] source_test.py
- [x] db_test.py
- [x] mumo_manager_test.py

### CODE REVIEW

**Comment by AnonFuncsAreAwesome** [comment]
I tested the rest of the modules (as far as I was able to test them) and fixed mumo_manager_test. Some logic with context callbacks was changed, caused one test to fail. Since samplecontext works fine that should be no problem.

@Kissaki I think this is ready to merge now, at least from my side.

**Comment by an0nfunc** [comment]
Anything new on this? I don't want to revisit my code in 5 years, again :)

**Comment by deluxghost** [comment]
I would like to start a smoke test in one or two months if nobody ever has time to move on this PR. I'm not sure if I can also review it entirely tho

**Comment by an0nfunc** [comment]
Any chance to get this merged?

**Comment by detrout** [comment]
I just ported MuMo to python3 as well (before remembering I should check to see if someone else did it) It would be nice to have this merged, Python 2 is past its end of life now.

**Comment by detrout** [comment]
I made a different stab at the conversion trying to maintain python 2 compatibility for now. (To give anyone using the current version some time to convert).

It's over here https://github.com/detrout/mumo/tree/python3-compat and I tried to break the changes into logically related chunks to make it easier to review.

**Comment by an0nfunc** [comment]
> I made a different stab at the conversion trying to maintain python 2 compatibility for now. (To give anyone using the current version some time to convert).

Since python2 is EOL since 11+ months now and most distros are going to get rid of if in their upcoming release if they have not done so already I do not think there is any more time needed, look at how long python3 is around now.

This should be long merged, it's mostly syntax change anyway.

**Comment by detrout** [comment]
I'd agree it'd be nice if the updated to Python3.

On my version I've gotten all the unit tests to pass which required implementing a stub customContextClass for MumoManager. There were also tests for announceContext and subscribeContextCallbacks which don't seem to exist and so I removed them.

I tested the bot with the seen, samplecontext, and chatimg modules enabled. (Though I had to do some porting to chatimg to get it to paste a black box. There's more to be done there).

**Comment by deluxghost** [comment]
I think there is no active maintainer for mumo now. Maybe one of you should join their team

**Comment by deluxghost** [comment]
@Kissaki any thoughts?

**Comment by an0nfunc** [comment]
Any news on this one?

> Anything new on this? I don't want to revisit my code in 5 years, again :)

I hope I was just joking here, right?

File: modules/bf2.py [hunk 1]
```diff
@@ -189,72 +190,74 @@ def update_state(self, server, oldstate, newstate):
...
         if oli and not nli:
-            log.debug("User '%s' (%d|%d) on server %d no longer linked", newstate.name, newstate.session, newstate.userid, sid)
+            log.debug("User '%s' (%d|%d) on server %d no longer linked", newstate.name, newstate.session,
+                      newstate.userid, sid)
             server.removeUserFromGroup(0, session, "bf2_linked")
-                
-        if newstate.channel >= 0 and newoldchannel != newstate.channel:
-            if ng == None:
+
+        if 0 <= newstate.channel != newoldchannel:
```
Comment (Kissaki) [inline modules/bf2.py:248]: thought: I find this double-sided comparison confusing; not very readable or clear.
One condition checks general validity of the value to be in the value range. The second checks for value change for this particular use case/section. So it mixes two concerns too.

File: modules/source/source_test.py [hunk 2]
```diff
@@ -29,89 +29,94 @@
...
     def addUserToGroup(self, cid, session, group):
         c = self._getChan(cid)
-        
+
         if session in c.groups:
             c.groups[session].add(group)
         else:
-            c.groups[session] = set([group])
-    
+            c.groups[session] = {group}
```
Comment (Kissaki) [inline modules/source/source_test.py:87]: This is confusing to me. It was a set, now it’s an object/dictionary?

File: mumo_manager_test.py [hunk 3]
```diff
@@ -30,155 +30,154 @@
...
-        mod.econnected.wait(timeout=1)
-        assert(mod.econnected.is_set())
-        man.announceContext(man.MAGIC_ALL, "contextCallMe", "server", "arg1", arg2 = "arg2")
-        mod.econtext.wait(timeout=1)
-        assert(mod.econtext.is_set())
-        man.announceDisconnected()
-        self.down(man, mod)
-    
+
+    # FIXME: Test ContextCallbacks correctly
```
Comment (Kissaki) [inline mumo_manager_test.py:154]: Note: Broken test; added FIXME note.

**Comment by Kissaki** [comment]
Thank you for implementing and suggesting these changes. Sorry for the delay and lack of response.

The biggest problem of this PR is it’s size. The changeset could have been a lot smaller if whitespace, formatting, and actual logical changes were seperate commits (or only the logical changes were the changeset). Having looked through the changes, I feel like 70% is unrelated whitespace and formatting changes.

I am not particularly involved with python, and have no test or productive setup for this/mumo, so I was hesitant to review this despite being tagged. The number of changes further overwhelmed and dejected me. ;-)

* trailing removal - fine by me
* some alignment changes - do not always make sense to me
* some formatting changes seem opinionated, and some definitely detrimental to conciseness

Either way I am fine with including them as-is as not to make more work.

It was said that there is probably no maintainer for mumo anymore. Evidently, that is the case. If you or anyone else would be willing to take that role, I’d gladly verify and accept.

The question of python2 backwards compatibility was risen. With its EOL I agree we no longer need that. Keeping it simpler on python3 is preferable now, for maintainability.

Even without me testing or using this myself, given that nobody else stepped up, I am willing to accept this changeset as-is. Moving forward is preferable to my perfectionism/reluctance here.

One test was commented out with a FIXME. I am fine with that so we can move forward. The tested functionality is context menu, so narrow compared to overall.

My only concern/confusion is with what looks like a type change to me on source engine PA gamestate unit test field initialization. But that is also a narrow, negligible use case compared to the overall usefulness of these changes.

So, I will accept and merge this as-is, in the interest of landing it without waiting for responses. If you are willing, feel free to follow up though. :)

**Review by Kissaki — APPROVED** [review]

File: modules/source/source_test.py (hunk 2 above)
Comment (an0nfunc) [reply modules/source/source_test.py:87]: Actually, that is [one way to create a set](https://docs.python.org/3/tutorial/datastructures.html#sets). So no chnage here, just a more 'pythonic' syntax.

File: modules/bf2.py (hunk 1 above)
Comment (an0nfunc) [reply modules/bf2.py:248]: I agree this is somewhat confusing, but the logic is actually the same. Most likely the whole if should be rewritten to make it more readable.

**Comment by an0nfunc** [comment]
Nice to see this finally merged. Even if not begin perfect - as you mentioned above - this still moved mumo forward and gives other contributors (and me) a chance to fix these smaller things in a python3 codebase. Thanks for you detailed explaining of your thought process and merging this.

File: modules/source/source_test.py (hunk 2 above)
Comment (Kissaki) [reply modules/source/source_test.py:87]: `{}` dict, `{1}` set, `{'a':1}` dict

ok…

**Pull request merged**